### PR TITLE
hackers delight 

### PR DESCRIPTION
**Pull request by AnotherAlexHere**

### CODE REVIEW

File: SSA/Projects/InstCombine/HackersDelight [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+theorem neg_eq_not_add_one {x : BitVec w} :
+    -x = ~~~x + 1 := by
+  exact BitVec.neg_eq_not_add x
+
+theorem neg_eq_neg_not_one {x : BitVec w} :
+    -x = ~~~(x - 1) := by
+  apply BitVec.eq_of_toNat_eq
+  sorry
+
+
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/HackersDelight]: Do not use two empty lines in sequence.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -37,19 +37,19 @@ lemma two_pow_pred_mod_eq_two_pred (h : w > 0): 2 ^ (w - 1) % 2 ^ w = 2 ^ (w - 1
...
-def ushr_and_distrib (a b c : BitVec w) :
+lemma ushr_and_distrib (a b c : BitVec w) :
     (a &&& b) >>> c = (a >>> c) &&& (b >>> c) := by
   simp only [HShiftRight.hShiftRight]
   ext
   simp
 
-def ushr_or_distrib (a b c : BitVec w) :
+lemma ushr_or_distrib (a b c : BitVec w) :
     (a ||| b) >>> c = (a >>> c) ||| (b >>> c) := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: These changes are unrelated and should not be included here.

**Comment by tobiasgrosser** [comment]
@AnotherAlexHere, please wait for at least one other person to review your changes.

**Comment by tobiasgrosser** [comment]
@AnotherAlexHere, if you could address my comments, that would be appreciated.

File: SSA/Projects/InstCombine/HackersDelight (hunk 1 above)
Comment (AnotherAlexHere) [reply SSA/Projects/InstCombine/HackersDelight]: Will fix this in next commit.

File: SSA/Projects/InstCombine/ForLean.lean (hunk 2 above)
Comment (AnotherAlexHere) [reply SSA/Projects/InstCombine/ForLean.lean]: Yes, you are right.

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: I want these changes to disappear. 😄

File: SSA/Projects/InstCombine/HackersDelight (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/HackersDelight]: And this line to disappear.

**Review by tobiasgrosser — APPROVED** [review]

File: SSA/Projects/InstCombine/HackersDelight [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import Init.Data.Nat.Bitwise.Lemmas
+
+namespace HackersDelight
+
+namespace Ch2Basics
+
+theorem neg_add_eq_neg_sub {x: BitVec w } :
+    -(x + y) = -x - y := by
+    ring
+
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/HackersDelight]:
```suggestion
  ring

```

Tactics are indentented by two whitespace only.

File: SSA/Projects/InstCombine/HackersDelight [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+theorem neg_eq_not_add_one {x : BitVec w} :
+    -x = ~~~x + 1 := by
+  exact BitVec.neg_eq_not_add x
+
+theorem not_eq_neg_sub_one {x : BitVec w} :
+    ~~~ x = -x - 1 := by
+  apply BitVec.eq_sub_iff_add_eq.mpr
+  rw[BitVec.neg_eq_not_add]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/HackersDelight]:
```suggestion
  rw [BitVec.neg_eq_not_add]
```

Whitespace between rw and `[`.

**Comment by tobiasgrosser** [comment]
@AnotherAlexHere, I am not sure we want to try to 'proof' these statements. I would suggest keeping the proofs where they go through and just leaving sorry everywhere else.

**Comment by tobiasgrosser** [comment]
The one thing that is important is that we actually build this file. Maybe @AtticusKuhn can help us to get it hooked to the build system?

File: SSA/Projects/InstCombine/ForLean.lean (hunk 2 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: Now these are back again. The last revert was not really needed, I think.

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: Try to work with @AtticusKuhn to clean up this PR. You may need some help with checking how to make this clean and ready.

**Review by tobiasgrosser — APPROVED** [review]

File: SSA/Projects/InstCombine/HackersDelight [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
+  ring
+
+theorem not_sub_eq_not_add {x y : BitVec w} :
+    ~~~ (x - y) = ~~~ x + y := by
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
+  ring
+
+theorem or_eq_not_or_sub_not {x y : BitVec w} :
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight:100]: Rename this theorem to `or_eq_and_not_sub`.

File: SSA/Projects/InstCombine/HackersDelight [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  sorry
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
+  ring
+
+theorem not_sub_eq_not_add {x y : BitVec w} :
+    ~~~ (x - y) = ~~~ x + y := by
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight:97]: Put a space after `rw`
`rw [not_eq_neg_sub_one, not_eq_neg_sub_one]`

File: SSA/Projects/InstCombine/HackersDelight [hunk 7]
```diff
@@ -0,0 +1,110 @@
+import Init.Data.BitVec.Basic
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight]: Rename this file to `HackersDelight.lean` so we get lean syntax highlighting

File: SSA/Projects/InstCombine/HackersDelight (hunk 5 above)
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight]:
```suggestion
theorem or_eq_and_not_sub {x y : BitVec w} :
```

File: SSA/Projects/InstCombine/HackersDelight [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    x &&& ~~~ y = (x ||| y) - y := by
+  sorry
+
+theorem and_not_eq_not_add {x y : BitVec w} :
+    x &&& ~~~ y = ~~~ x + y := by
+  sorry
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight]: Put a space after rw
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 9]
```diff
@@ -1,14 +1,110 @@
...
+  exact BitVec.neg_eq_not_add x
+
+theorem not_eq_neg_sub_one {x : BitVec w} :
+    ~~~ x = -x - 1 := by
+  apply BitVec.eq_sub_iff_add_eq.mpr
+  rw [BitVec.neg_eq_not_add]
+
+theorem neg_eq_neg_not_one {x : BitVec w} :
+    -x = ~~~(x - 1) := by
+  rw[not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 10]
```diff
@@ -1,14 +1,110 @@
...
+  ring
+
+theorem not_not {x : BitVec w} :
+    ~~~ ~~~ x = x := by
+  rw[not_eq_neg_sub_one,not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one,not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 11]
```diff
@@ -1,14 +1,110 @@
...
+  ring
+
+theorem neg_not_eq_add_one {x : BitVec w} :
+    - ~~~ x = x + 1 := by
+  rw[BitVec.neg_eq_not_add, not_not]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [BitVec.neg_eq_not_add, not_not]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 12]
```diff
@@ -1,14 +1,110 @@
...
+
+theorem not_neg_eq_sub_one {x : BitVec w} :
+    ~~~ (-x) = x - 1 := by
+  rw[not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 13]
```diff
@@ -1,14 +1,110 @@
...
+  ring
+
+theorem add_eq_sub_not_sub_one {x y : BitVec w} :
+    x + y = x - ~~~ y - 1 := by
+  rw[not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 14]
```diff
@@ -1,14 +1,110 @@
...
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
+  ring
+
+theorem not_sub_eq_not_add {x y : BitVec w} :
+    ~~~ (x - y) = ~~~ x + y := by
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
+  ring
+
+theorem or_eq_not_or_sub_not {x y : BitVec w} :
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
theorem or_eq_and_not_add {x y : BitVec w} :
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 15]
```diff
@@ -1,14 +1,110 @@
...
+    x &&& ~~~ y = (x ||| y) - y := by
+  sorry
+
+theorem and_not_eq_not_add {x y : BitVec w} :
+    x &&& ~~~ y = ~~~ x + y := by
+  sorry
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 16]
```diff
@@ -1,14 +1,110 @@
...
+    x + y = (x ||| y) + (x &&& y) := by
+  sorry
+
+theorem add_eq_mul_or_neg_xor {x y : BitVec w} :
+    x + y = 2 * (x ||| y) - (x ^^^ y) := by
+  sorry
+
+theorem sub_eq_add_not_add_one {x y : BitVec w} :
+    x - y = x + ~~~ y + 1 := by
+  rw[not_eq_neg_sub_one]
```
Comment (AnotherAlexHere) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one]
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 17]
```diff
@@ -1,14 +1,110 @@
...
+  sorry
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
+  ring
+
+theorem not_sub_eq_not_add {x y : BitVec w} :
+    ~~~ (x - y) = ~~~ x + y := by
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
```
Comment (AnotherAlexHere) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
  rw [not_eq_neg_sub_one, not_eq_neg_sub_one]
```

**Comment by AtticusKuhn** [comment]
> The one thing that is important is that we actually build this file. Maybe @AtticusKuhn can help us to get it hooked to the build system?

I added the following job to the Github CI


```yaml
      - name: Compile Hacker's Delight Theorems 🧮
        run: |
          lake -R exe cache get # download cache of mathlib docs.
          lake -R build SSA.Projects.InstCombine.HackersDelight # compile and check the Hacker's Delight theorems
```

Hopefully, this should work.

File: .github/workflows/ci.yml [hunk 18]
```diff
@@ -36,6 +36,12 @@ jobs:
           lake -R exe cache get # download cache of mathlib docs.
           lake -R build SSA.Experimental.Bits.Decide SSA.Experimental.Bits.Fast.Decide
 
+      - name: Compile Hacker's Delight Theorems 🧮
+        run: |
+          lake -R exe cache get # download cache of mathlib docs.
+          lake -R build SSA.Projects.InstCombine.HackersDelight # compile and check the Hacker's Delight theorems
+
+
       - name: Compile Alive Examples
```
Comment (tobiasgrosser) [inline .github/workflows/ci.yml]:
```suggestion

      - name: Compile Alive Examples
```

**Pull request merged**